sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/namedExpressions.scala

```diff
@@ -415,6 +415,25 @@ case class PrettyAttribute(
   override def nullable: Boolean = true
 }
 
+/**
+ * An expression that has to be resolved against a scope of resolved attributes.
+ */
+case class ScopedExpression(expr: Expression, scope: Seq[Attribute])
```

Comment:
This is not a named expression. shall we move it to a new file, with the two new analyzer rules?

Reply:
As in bb1468ad876ccc9c77c3af25945c35301a5f9114?